consensus/merge/merge.go

```diff
@@ -159,20 +160,53 @@ func (s *Merge) Finalize(config *chain.Config, header *types.Header, state *stat
 		}
 	}
 
-	return txs, r, nil
+	var rs types.Requests
+	if config.IsPrague(header.Time) {
+		rs = make(types.Requests, 0)
+		allLogs := types.Logs{}
+		for _, rec := range receipts {
+			allLogs = append(allLogs, rec.Logs...)
+		}
+		ds, err := types.ParseDepositLogs(allLogs, config.DepositContract)
+		rs = append(rs, ds...)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("error: could not parse requests logs: %v", err)
+		}
+
+		rs = append(rs, misc.DequeueWithdrawalRequests7002(syscall)...)
+		if requests != nil || header.RequestsRoot != nil {
+			rh := types.DeriveSha(rs)
+			if *header.RequestsRoot != rh {
+				return nil, nil, nil, fmt.Errorf("error: invalid requests root hash in header, expected: %v, got :%v", header.RequestsRoot, rh)
+			}
+			sds := requests.Deposits()
+			if !reflect.DeepEqual(sds, ds.Deposits()) {
```

Comment:
We probably need a similar check for withdrawal requests - see https://discord.com/channels/595666850260713488/892088344438255616/1247841586147229696

Reply:
Withdrawal request validations are done at the smart contract itself, so `dequeue` would get out the exact set of withdrawals.
Deposits are obtained from logs, which happens outside of EVM and in the above lines, and must be checked.
However, this check is already embedded with checking the `requestsRoot` as there is no scenario in which a different set of deposits passed in would lead to the same `requestsRoot`.

Reply:
I still don't understand why there should be a difference between deposit and withdrawal requests in terms of checks. At the end of a block expected deposit requests are calculated by `ParseDepositLogs(logs)`, while withdrawal requests are calculated by `DequeueWithdrawalRequests7002(syscall)`. If we rely solely on the `*header.RequestsRoot != rh` check for withdrawal requests, then we should do the same for deposit ones and remove the `!reflect.DeepEqual(sds, ds.Deposits())` check.

Reply:
Yes, I will add the `DeepEqual` check for `withdrawal_requests` in a later PR - under the assumption that there can be other request types in future.
Alternatively, until there is another type of request introduced, we can remove this check altogether and rely on `block.HashCheck()` combined with the root check above